### PR TITLE
Dyno: ensure sufficient lifetime of data in reported error messages

### PR DESCRIPTION
**Pull request by DanilaFe**
Closes https://github.com/Cray/chapel-private/issues/5854.

This PR resolves a Dyno memory issue that was originally discovered in code that:

1. Parses to the same AST
2. Produces different errors

caused segmentation faults. This most commonly happened when there was a parse error in the program text, and the parser dropped the "unrecognized" portion of the file, resulting in the same tree both times. @jabraham17's original reproducer involved changing the contents of an invalid identifier `foo` to `fooo` or `fo` (this identifier didn't make it into the parse tree, but did appear in the error message)

```Chapel
use IO;

module R {
  proc bar() {
foo
  }
}
````

The bug is actually more broad, and has to do with the query system's re-computation model and the lifetime of data that goes into error messages. The key idea here is that we avoid re-running queries whose dependencies produced the same result as before, __even if the dependencies computed that result differently__. For instance, if a query computes the number of functions in a file, even if the user re-orders the functions, the result of the query doesn't change, and dependent queries are not recomputed.

The case in the previous paragraph is easy for integers, but more difficult for arbitrary data. Things boil down to having to perform a deep comparison of the old and new query results, which could include heap-allocated memory (e.g., the parser allocates memory for AST nodes). If the comparison succeeds, we know that the query output hasn't changed, so there shouldn't be a need to re-run parent queries. However, it's allowed for the parent queries to rely on the exact memory locations etc. of the dependency (e.g., a dependent query could compute an `AstNode* -> Type` map), so __if we want to soundly avoid re-computing dependent queries, we need to keep the _old_ result of the dependency query.__ We thus throw away the new (and equivalent) result, and keep the old one.

However, as is the case with Jade's test case, equivalent results need not be produced in the same way. Each time `foo` is changed to something else, the query computing the AST of `module R` emits a different error message, in terms of the new AST. In general, __errors issued while recomputing a query (and seeing if it changed) reference newly allocated data__. More generally, the code computing the error message can differ arbitrarily, and we could allocate lots of intermediate data 

But then, since we throw away the new result in favor of the old, we end up with errors that contained deallocated memory, leading to segmentation faults.

It seems reasonable to maintain the latest errors in the query, so that the _actual_ error emitted can be queried, instead of the-first-error-message-that-led-to-this-result. However, this means that if we emitted error messages, we can't throw away the new-but-equivalent result. Dyno already has a mechanism for storing old results for the purposes of memory lifetimes, so my original PR made the following change:

```diff
- if (initialResult || changed==false) {
+ if ((initialResult || changed==false) && !r->errors.empty()) {
    // no need to save old result
  } else {
    queryMap->oldResults.push_back(std::move(result));
  }
```

However, `oldResults` gets cleared when garbage collection is performed. This leaves possible the case when:

1. The query is initially computed
2. The query is re-computed, producing the same result but issuing new errors
  * So, we keep the new memory for the purpose of the errors in `oldResults`
3. Garbage collection is performed, clearing `oldResults`
4. The query is re-run, but its dependencies haven't changed, so we issue the stored errors, which depend on the now-cleared `oldResults`.

The bottom line is that error messages must survive garbage collection just like query results do, which means the data they depend on must too. To implement this, this PR adds an index into the `oldResults` list that should be kept alive past garbage collection. There can only be one such index per `QueryMapResult` since the result only stores the latest errors.

Reviewed by @mppf -- thanks!

## Testing
- [x] Jade's example in CLS no longer crashes
- [x] Jade's C++-based reproducer no longer crashes, and is included with the dyno tests, which all still pass.

### CODE REVIEW

File: frontend/include/chpl/framework/query-impl.h [hunk 1]
```diff
@@ -463,7 +463,15 @@ Context::updateResultForQueryMapR(QueryMap<ResultType, ArgTs...>* queryMap,
...
   // no need to save old result 1st time running this query
   // no need to save new result when old one is used due to no changes
-  if (initialResult || changed==false) {
+  //
+  // but, if we issued errors, even if 'changed' is false, the errors may
+  // contain newly-allocated data from the result, so keep that data.
+  if (initialResult) {
+    // no need to save old result
+  } else if (!r->errors.empty()) {
+    r->oldResultForErrorContents = queryMap->oldResults.size();
```
Comment (mppf) [inline frontend/include/chpl/framework/query-impl.h:472]: let's add a comment before this line identifying this as a workaround and indicating the problem it's solving (or link to the issue).

**Review by mppf — APPROVED** [review]
This is OK with me as a near-term fix but I'd like to see some comments indicating it's a TODO to find a better resolution.

One thing that worries me about this approach is that the error-message printing code might do things like compare pointers between uAST nodes. Is it possible we could get in a situation where it would compare uAST pointers across different generations and get a surprising `false` result?

If I am understanding correctly, when doing a series of incremental compilations, the queries that return error messages will be kept around for 1 generation longer, assuming they are attempted again. But these will always be hanging around in some form. It might not be worth worrying about but it feels a bit like this change just makes such queries never get freed. That said, I think it's reasonable to focus primarily on code without errors, in terms of memory/performance.

Anyway, all that said, I'm happy enough with this going in if you can improve the comments to make it clearer that saving this data is a work-around & at some point we might have a more principled way to handle it.

**Comment by mppf** [comment]
> In my opinion, all of this makes this PR a sound long-term fix, in the spirit of the query system. However, if you believe we ought to find a different solution, I am fine to treat it as a workaround and update the comments as such.

It might well be the best solution in the long term. Or it might not. I think it's worth putting a comment that says that there might be a better solution that we haven't thought of yet.

> I don't see how. The main reason is twofold:

Yeah, but there could be other things going on. E.g., there could be an application using the frontend library that is looking for error messages mentioning a particular uAST node. It might parse & save that uAST node pointer in a global & then traverse the error messages. It is distinctly strange that it might process a copy of the uAST that isn't the one currently loaded. Of course, you could say "don't do that", but what I described would be valid according to the query framework design, provided the global is updated on each revision.

**Comment by DanilaFe** [comment]
> Yeah, but there could be other things going on. E.g., there could be an application using the frontend library that is looking for error messages mentioning a particular uAST node. 

Thank you, I hadn't considered that.
